### DISCUSSION TITLE
API Documentation

### DISCUSSION THREAD

**Discussion by JoshClose** [statiqdev/Discussions#62] [Q&A]
I'm writing an API documentation generator. It's currently using [NuDoq](https://github.com/devlooped/NuDoq) to parse the assembly.

What would be the best way to inject this into Static.Web processing? It will be generating content, metadata, and possibly more.

There are no input files needed. It will just need to read some config values, like where to find the assembly it's generating the docs from.

**Comment by daveaglick** [comment]
Cool! Just FYI, [Statiq will eventually have in-the-box API documentation support via Statiq Docs](https://github.com/statiqdev/Statiq.Docs) but I'm all about keeping stuff open so lots of different techniques can be used as needed (just wanted to make sure you knew it was coming).

My first instinct would be to add the NuDoq code to a module and then add that module to a new pipeline (let's call it `Api`). Then add the `Api` pipeline as a dependency of the `Input` pipeline:

```csharp
.ModifyPipeline(nameof(Pipelines.Inputs), x => x.Dependencies.Add(nameof(Api)))
```

You could have the module generate a document-per-symbol, or maybe a document-per-class. Then, since it's a dependency of the `Inputs` pipeline you can iterate over those documents in other pages and layouts or create [archives](https://statiq.dev/web/content-and-data/archives) from them.

Does that make sense for a starting point?

**Reply by JoshClose** [reply]
Do you have an idea of how long until Statiq Docs will be ready?

**Reply by daveaglick** [reply]
My pals on the [Cake](https://cakebuild.net/) team and I are currently getting a head start by trying to port the API documentation there to Statiq (it's based Wyam which was a precursor to Statiq). With their help (especially @bjorkstromm - credit where's it's due, he's awesome) the effort is hopefully getting a good jump-start. Porting the theme is going to be the hardest part, but I'm hoping even that isn't too bad once we really start rolling.

TL;DR: I'm not sure exactly when it'll be ready, but work is actively underway and it's quickly becoming a primary focus. Maybe a few weeks, maybe a bit longer. But certainly not another year or anything like that.

**Reply by JoshClose** [reply]
OK. I think I'll wait for you to do it since you know the system.

For your reference, I need to generate documents that look like this and contain this info. Including a TOC for navigation. https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.hub-1?view=aspnetcore-5.0

**Reply by daveaglick** [reply]
Hopefully it'll be close - the kind of API documentation that Wyam produces looks like https://wyam.io/api/, https://cakebuild.net/api/, and https://www.reactiveui.net/api/ so Statiq will likely look pretty similar (at least initially).

**Reply by daveaglick** [reply]
As with Statiq Web though, the data and pipelines will all be part of Statiq Docs but the theme and how that data is rendered will be separate. That should make it easy enough to swap out a different theme or manipulate the default one.

**Reply by JoshClose** [reply]
That should work.

**Reply by JoshClose** [reply]
Will there be a way to merge example code and other comments into the API docs?

**Reply by bjorkstromm** [reply]
It's happening 😄

https://twitter.com/bjorkstromm/status/1367371910313574401?s=20

![Statiq Docs](https://user-images.githubusercontent.com/7863439/109925418-65e9ef00-7cca-11eb-8147-40f6e6ffe6da.gif)

**Reply by bjorkstromm** [reply]
> 
> 
> Will there be a way to merge example code and other comments into the API docs?

How do you mean? Like merge a "normal" comment with XML Doc? E.g.

```csharp
/// <summary>
/// This is a method.
/// </summary>
/// <param name="foo">This is a param</param>
public void Foo(string foo)
{
    // This is another comment that I want included in docs?
    _foo = foo;
}
```

**Reply by JoshClose** [reply]
Probably like a sidecar file or something. I know DocFX does this. I don't remember off hand how they achieve it, but I would assume Statiq would do it in a much better way anyways. 🙂

**Reply by bjorkstromm** [reply]
Can you point me to an example on what DocFX does? It's probably like you say, if DocFX does it, it shouldn't be a problem for Statiq either.

**Reply by daveaglick** [reply]
There's actually a standard for these files. Basically you stick the same XML doc comments you'd add to the symbol in a stand-alone XML file and then use `<include>` to reference them: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/include. Wyam used to support this (though there were some problems and I think it might have gotten dropped at some point) and I've got every intention of Statiq supporting it.

**Reply by JoshClose** [reply]
> When the C# compiler sees the <include> tag, it searches for documentation comments in xml_include_tag.doc **instead** of the current source file.

I just tried with a `<summary>` in source and all the `<param>`s in an include file and the xml file that was built contained both. It seems to be merging them. I'll probably continue to put most of the docs in code, then larger blocks like `<example>` could go in another file.

I believe this also means there is no extra work on your part. It's just built in.

Make sure to handle user defined tags when parsing the docs also.

This doc has a much better example. https://docs.microsoft.com/en-us/dotnet/csharp/codedoc#include

**Reply by daveaglick** [reply]
> I believe this also means there is no extra work on your part. It's just built in.

I wish! :)

The XML documentation support is...interesting. Roslyn (which is what Statiq generally uses under the hood for code analysis) supports most of the conventions and standards that are detailed in the MS docs, but not necessarily out of the box. A few of the features like these external XML files require a little extra handholding and configuration. And on top of that, Visual Studio gets fancy with some of the more exotic pseudo-standards. For example, ever noticed that `<inheritdoc>` works in Visual Studio? That's not actually part of Roslyn but something Visual Studio added on top of it.

**Reply by JoshClose** [reply]
`<include>` also works with in VS which is nice.

I guess I assumed you would be opening the Library.xml file that is generated and using that.
